### PR TITLE
Add partition and memengine cusomer catalog request

### PR DESCRIPTION
**Pull request by nnsgmsone**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3931

## What this PR does / why we need it:
cn: add parition
memengine: cusomer  catalog request
memengine: cusomer table request

### CODE REVIEW

**Review by daviszhen — APPROVED** [review]

File: pkg/catalog/catalog.go [hunk 1]
```diff
@@ -45,3 +49,252 @@ func newAttributeDef(name string, typ types.Type, isPrimary bool) engine.TableDe
...
+// consume a set of entries and return a command and the remaining entries
+func ParseEntryList(es []*api.Entry) (any, []*api.Entry, error) {
+	if len(es) == 0 {
+		return nil, nil, nil
+	}
+	e := es[0]
+	if e.DatabaseId != MO_CATALOG_ID {
+		return e, es[1:], nil
+	}
+	switch e.TableId {
```
Comment (fengttt) [inline pkg/catalog/catalog.go:62]: I remember there are 3 tables?  Only two left now?

File: pkg/catalog/catalog.go [hunk 2]
```diff
@@ -45,3 +49,252 @@ func newAttributeDef(name string, typ types.Type, isPrimary bool) engine.TableDe
...
+		return cmds, es[:idx], nil
+	default:
+		return e, es[1:], nil
+	}
+}
+
+func genCreateDatabases(rows [][]any) []CreateDatabase {
+	cmds := make([]CreateDatabase, len(rows))
+	for i, row := range rows {
+		cmds[i].Name = string(row[1].([]byte))
```
Comment (fengttt) [inline pkg/catalog/catalog.go:99]: All these strange numbers ...

Get a const.

Comment (nnsgmsone) [reply pkg/catalog/catalog.go:99]: ok

File: pkg/catalog/catalog.go (hunk 1 above)
Comment (nnsgmsone) [reply pkg/catalog/catalog.go:62]: 3,  mo_database, mo_tables, mo_columns.

**Review by fengttt — COMMENTED** [review]
Fix your constants.

**Review by yingfeng — APPROVED** [review]

File: pkg/catalog/catalog.go (hunk 1 above)
Comment (fengttt) [reply pkg/catalog/catalog.go:62]: Did you just handle 2?    Let's handle 3 explicitly and default case, panic with an error.

**Review by reusee — APPROVED** [review]

File: pkg/catalog/catalog.go (hunk 1 above)
Comment (nnsgmsone) [reply pkg/catalog/catalog.go:62]: I handled three of them ， create database generates tuples written to mo_database, and create database generates tuples written to mo_tables and mo_columns.
 consumes logtail does not care whether the specific table is a catalog or not,，https://github.com/matrixorigin/matrixone/blob/main/pkg/vm/engine/disttae/logtail.go#L81。

Comment (nnsgmsone) [reply pkg/catalog/catalog.go:62]: Then when I read the catalog information again, I pass in its specific id. for example: https://github.com/matrixorigin/matrixone/blob/5583ac15a3f690b3ce15815702174d003b394221/pkg/vm/engine/disttae/txn.go#L28

**Review by zhangxu19830126 — APPROVED** [review]

File: pkg/catalog/catalog.go (hunk 1 above)
Comment (fengttt) [reply pkg/catalog/catalog.go:62]: If mo_columns should never reach here, panic an internal error as well.

**Review by aunjgr — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Pull request merged**